examples/tut-6-rpc/client-go/index.go

```diff
@@ -0,0 +1,214 @@
+package main
+
+import (
+  "fmt"
+  "log"
+  "os"
+  "reflect"
+  "strings"
+  "sync"
+  "time"
+  "encoding/json"
+  "io/ioutil"
+  "math/rand"
+  "net/http"
+)
+
+
+type account   = float64
+type contract  = float64
+type bigNumber = map[string]interface {}
+
+
+func dieIf(err error) {
+  if err != nil {
+    log.Fatal(err)
+    os.Exit(1)
+  }
+}
+
+
+func please(a interface{}, err error) interface{} {
+  dieIf(err)
+  return a
+}
+
+
+func mkRpc() (func(string, ...interface{}) interface{},
+              func(string, contract, map[string]interface{})) {
+
+  proto := "http"
+  host  := os.Getenv("REACH_RPC_SERVER")
+  port  := os.Getenv("REACH_RPC_PORT")
+
+  // TODO wait for server to become available
+
+  rpc := func(m string, args ...interface{}) (interface{}) {
+    var ans   interface{}
+    var jargs string
+
+    if args == nil {
+      jargs = "[]"
+    } else {
+      jargs = string(please(json.Marshal(args)).([]byte))
+    }
+
+    lab  := fmt.Sprintf("RPC %s %s\n", m, jargs)
+    fmt.Printf(lab)
+
+    uri  := fmt.Sprintf("%s://%s:%s%s", proto, host, port, m)
+    res  := please(http.Post(uri, "application/json", strings.NewReader(jargs))).(*http.Response)
+    body := please(ioutil.ReadAll(res.Body)).([]byte)
+
+    res.Body.Close();
+    dieIf(json.Unmarshal(body, &ans))
+
+    fmt.Printf("%s ==> %s\n", lab, string(body))
+
+    return ans
+  }
+
+  rpcCallbacks := func(m string, arg contract, cbacks map[string]interface{}) {
+    var ans interface{}
+
+    vals  := make(map[string]interface{})
+    meths := make(map[string]bool)
+
+    for key, val := range cbacks {
+      if reflect.ValueOf(val).Kind() == reflect.Func {
+        meths[key] = true
+      } else {
+        vals[key]  = val
+      }
+    }
+
+    p := rpc(m, arg, vals, meths).(map[string]interface{})
+
+    for true {
+      if p["t"] == "Done" {
+        break
+
+      } else if p["t"] == "Kont" {
+
+        // Apply interact method to supplied inputs
+        if _, exists := meths[p["m"].(string)]; exists {
+          cb    := reflect.ValueOf(cbacks[p["m"].(string)])
+          args  := p["args"].([]interface{})
+          vargs := make([]reflect.Value, len(args))
+
+          for i, a := range args {
+            vargs[i] = reflect.ValueOf(a)
+          }
+
+          ans = nil
+          for _, a := range cb.Call(vargs) {
+            ans = a.Interface()
+            break
+          }
+
+          p = rpc("/kont", p["kid"], ans).(map[string]interface{})
+
+        // Respond with simple field value
```

Comment:
I don't understand this... weren't these things transmitted in `vals`? What error did you have before this... it seems like a problem somewhere else

Reply:
This is one of the things I was alluding to when I said I was still poking at it - I think you're right and I can remove this part.

Trying to explain why it's here wouldn't make much sense without context that's stale now :-)

Reply:
:)

Reply:
Rejiggered in d2164e4.